src/workerd/api/sockets.c++

```diff
@@ -128,4 +132,30 @@ void Socket::handleProxyStatus(
   result.markAsHandled();
 }
 
+jsg::Promise<void> Socket::closeWriteSide(jsg::Lock& js) {
+  // When `allowHalfOpen` is set to true then we do not automatically close the write side on EOF.
+  // The default value for `allowHalfOpen` is also false.
+  KJ_IF_MAYBE(opts, options) {
+    if (opts->allowHalfOpen) {
+      return js.resolvedPromise();
+    }
+  }
+
+  // We want to close the socket, but only after its WritableStream has been flushed. We do this
+  // below by calling `close` on the WritableStream which ensures that any data pending on it
+  // is flushed. Then once the `close` either completes or fails we can be sure that any data has
+  // been flushed.
+  return writable->getController().close(js).then(js, [this](jsg::Lock& js) {
+    closeFulfiller->resolver.resolve();
+  }, [this](jsg::Lock& js, jsg::Value err) {
+    // A failure to close the WritableStream can indicate one of these things:
+    //   * The WritableStream hasn't been attached.
+    //   * The WritableStream has already been released or closed.
+    //
+    // We only want to ensure that the writable stream is flushed before closing the socket.
+    // With the above we are certain that it was flushed, so we are safe to close.
+    closeFulfiller->resolver.resolve();
+  });
```

Comment:
I believe you can simplify to something like...

```cpp
return writable->getController().close(js).catch_(js, [](jsg::Lock& js) {
    // A failure to close the WritableStream can indicate one of these things:
    //   * The WritableStream hasn't been attached.
    //   * The WritableStream has already been released or closed.
    //
    // We only want to ensure that the writable stream is flushed before closing the socket.
    // With the above we are certain that it was flushed, so we are safe to close.
}).then(js, JSG_VISITABLE_LAMBDA((ref=JSG_THIS), (ref), (jsg::Lock& js) {
  ref->closeFulfiller->resolver.resolve();
}));
```

Using the `JSG_VISITABLE_LAMBDA` and capturing the `JSG_THIS` will ensure that the `Socket` instance safely remains alive while waiting for the continuation.